Add relative lock time type

tcharding opened the pull request:
Patches 1 and 2 are preparation. Patch 3 adds a new `relative::LockTime` type.

Please see https://github.com/rust-bitcoin/rust-miniscript/pull/455 for demo of use of the new API in miniscript.

### Some links to help during review
- https://medium.com/summa-technology/bitcoins-time-locks-27e0c362d7a1
- https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki
- https://github.com/bitcoin/bips/blob/master/bip-0112.mediawiki

nlanson reviewed (commented):
Quick review.

nlanson commented on src/blockdata/transaction.rs:
```diff
@@ -380,6 +382,47 @@ impl Sequence {
+    /// Returns the low 16 bits from sequence number.
+    ///
+    /// BIP-68 only uses the low 16 bits for relative lock value.
+    pub(crate) fn value(&self) -> u16 {
```
or `low_u16()`

tcharding replied:
I like `low_16`, cheers.

tcharding commented on src/blockdata/transaction.rs:
```diff
@@ -380,6 +382,47 @@ impl Sequence {
+    /// Returns the time value used to satisfy a [`relative::LockTime`] if this is a time-based
+    /// sequence locking value.
+    pub(crate) fn time_value(&self) -> Option<u16> {
+        if self.is_height_locked() {
```
Correct! Thanks man.

tcharding commented:
Implemented suggestions above and rebased.

tcharding commented:
Rebased on top of 'disable fuzzer' patch, slightly improved docs.

apoelstra commented on src/blockdata/locktime/relative.rs:
```diff
@@ -0,0 +1,155 @@
+
+        match *self {
+            Blocks(height) => height.0 <= n,
+            Time(_) => false,
```
In 2e6a8aefc048c28f71d8a4efb5f1d2766354c9f2:

I think we shoud return `true` by default. Conceptually a height-based timelock has no constraint on time.

tcharding replied:
Done.

apoelstra commented on src/blockdata/locktime/relative.rs:
````diff
@@ -0,0 +1,155 @@
+    /// assert!(lock.is_satisfied_by(Sequence::from_height(height+1)));
+    /// ```
+    // FIXME: Have I got the terminology around the correct way? Does self satisfy n or n satisfy self?
+    pub fn is_satisfied_by(&self, n: Sequence) -> bool {
````
Oh, if you keep the existing `false` returns (see below) then `&&` should be replaced by `||`.

tcharding replied:
I'm still not fully understanding where the height and time values would come from to pass into this new function. Would a user of this function have access to the chain and be expected to be able to compute the current height and time?  I've based this (and the `absolute::LockTime` work) off of the usage in `miniscript`, is it that `miniscript` is not a normal user? How would we implement `check_older` in `miniscript` if we cannot use a `Sequence` number to satisfy the relative lock time?

apoelstra replied:
Ah, I understand the question now. So, there are two pieces to locktime validation in Bitcoin:

1. The `CHECKSEQUENCEVERIFY` (and `CHECKLOCKTIMEVERIFY`) opcodes, which check the sequence/locktime values in a transaction against some "test" sequence/locktime value, and decides whether or not it passes. This is what rust-miniscript uses when reasoning about `check_older`/`check_after`.
2. **Separately** the Bitcoin blockchain checks whether the sequence/locktime values actually indicate a valid locktime, based on the state of the chain. This is what I think "normal users" are thinking of when they work with sequence/locktime values.

I would suggest that a "normal user" is more likely to be in situation (2), whereas situation (1) corresponds to Script interpretation. In this API we should somehow support both usecases somehow

apoelstra replied:
I think that my suggested API, in which `is_satisfied_by` takes both a height and time, would serve both cases.

In case (1), rust-miniscript will see `check_older(N)`, decide whether to interpret `N` as a height or time, and then use `is_satisfied_by_height`/`is_satisfied_by_time` as appropriate. In case (2) the user has both the current height and current time, and can use `is_satisfied_by`.

Miniscript *additionally* needs a way to serialize a height/time as a "sequence number" to embed into Script, and to parse out such a value.

tcharding replied:
Thanks man! Your suggested API works quite nicely, here is `check_older` from `miniscript`
```rust
     match self.to_relative_lock_time() {
            None => false,
            Some(seq) => match (n, seq) {
                (Blocks(_), Blocks(h)) => n.is_satisfied_by_height(h.value()),
                (Time(_), Time(t)) => n.is_satisfied_by_time(t.value()),
                _ => false, // Not the same units.
            }
        }

```

tcharding replied:
Although it could equally have been written without using the `is_satisfied_by_*` methods
```rust
        match self.to_relative_lock_time() {
            None => false,
            Some(seq) => match (n, seq) {
                (Blocks(n), Blocks(lock_time)) => n.value() <= lock_time.value(),
                (Time(n), Time(lock_time)) => n.value() <= lock_time.value(),
                _ => false, // Not the same units.
            }
        }
```

I was only matching against `self` and tests were failing until I matched against both `self` and `n`, so the API is not bullet proof.

apoelstra replied:
It probably can't be helped, unless you pull that entire block of code into its own API function here :)

Kixunil replied:
Idea how to resolve having to come up with different names:

```rust
// maybe this doesn't have to be sealed, IDK
pub trait SatisfiesRelativeLockTime: sealed::SatisfiesRelativeLockTime {
    fn satisfies_relative_lock_time(self, lock_time: relative::LockTime) -> bool;
}

impl SatisfiesRelativeLockTime for Sequence { /* ... */ }
impl SatisfiesRelativeLockTime for (Height, Time) { /* ... */ }
// impl SatisfiesRelativeLockTime for (Time, Height) { /* ... */ } // for convenience so people don't have to remember the order?

impl relative::LockTime {
    pub fn is_satisfied_by<T: SatisfiesRelativeLockTime>(self, satisified_by: T) -> bool {
        satisfied_by.satisfies_relative_lock_time(self)
    }
}
```

tcharding replied:
I like the trait idea. Implementing for `Sequence` goes against what @apoelstra says above (quoted below)

> I don't think it makes sense for a locktime to be satisfied by a sequence no, or vice-versa. Rather, I would expect a sequence number to be convertable to a relative::Locktime, and then this would be satisfied by a blockheight and/or timestamp.

> I don't think it makes sense for a locktime to be satisfied by a sequence no, or vice-versa. Rather, I would expect a sequence number to be convertable to a relative::Locktime, and then this would be satisfied by a blockheight and/or timestamp.

If `Sequence` disables lock time it is not possible to construct a `relative::LockTime` from the sequence so this concern is addressed. I think we should implement for `relative::LockTime`, `(Height, Time)`, and   `(Time, Height)`.

tcharding replied:
I implemented for `Height` and `Time` as well as for the tuple of both.

Kixunil replied:
> If `Sequence` disables lock time it is not possible to construct a `relative::LockTime` from the sequence so this concern is addressed.

I still didn't like the possibility of manual check but now I think maybe that's too much paranoia.

Kixunil replied:
Oh, crap, I now realized it should've been the other way around. It makes sense to check if sequence is satisfied by `(Height, Time)`|`RelativeLockTime` but not that much for relative lock time.

apoelstra commented:
Done reviewing 2e6a8aefc048c28f71d8a4efb5f1d2766354c9f2.

tcharding commented:
I've updated https://github.com/rust-bitcoin/rust-miniscript/pull/455 as well to demo current state of this PR.

tcharding commented:
Implemented the `LockTime::is_satisfied_by` method using trait suggestion by @Kixunil. As a separate patch we remove the error used by `Sequence` and use the new one from the `relative` module (which I cut'n pasted). Can squash if requested.

tcharding commented:
Note that now in miniscript we have the super sexy
```rust
impl<Pk: MiniscriptKey + ToPublicKey> Satisfier<Pk> for Sequence {
    fn check_older(&self, n: relative::LockTime) -> bool {
        match self.to_relative_lock_time() {
            None => false,
            Some(lock_time) => n.is_satisfied_by(lock_time),
        }
    }
}
```

Kixunil commented:
Damn, now that I think about it, silently returning `false` in case of different units is probably a footgun. `miniscript` may need it and even use it correctly but it doesn't seem wise to provide such a footgun to everyone. We could return `Result` with `Infallible` in case of `(Height, Time)` but it's starting to get complicated.

apoelstra commented on src/blockdata/locktime/relative.rs:
```diff
@@ -0,0 +1,281 @@
+            Ok(Time::from_512_second_intervals(interval))
+        } else {
+            Err(Error::IntegerOverflow(seconds))
+        }
```
In fb5abb43711a9f39e62f0e7b25dda8f85b32e98b:

nit: this whole functtion could just call `from_seconds_floor(seconds + 511)`.

tcharding replied:
Realized while implementing this that doing so would put the modified `seconds` into the error. I left as is for now. If we want to change we should do it here and in `transaction.rs` as well.

apoelstra replied:
Ok, that's fair.

apoelstra commented on src/blockdata/transaction.rs:
```diff
@@ -381,6 +382,31 @@ impl Sequence {
+    ///
+    /// BIP-68 only uses the low 16 bits for relative lock value.
+    fn low_u16(&self) -> u16 {
+        (self.0 & 0x0000_FFFF) as u16
```
In fb5abb43711a9f39e62f0e7b25dda8f85b32e98b:

The masking seems unnecessary.

tcharding replied:
I did it on purpose to make the `as u16` explicitly safe because of the talk recently (@dpc and @Kixunil) about not use `as`. Can remove it (or leave it and document why).

apoelstra replied:
I have a mild preference for removing it. I think the doccomment already justifies the loss of bits. The masking adds to the cognitive load.

tcharding replied:
Sure thing, thanks.

apoelstra commented:
Just a couple of nits.

I see what @Kixunil is saying that silently returning `false` (actually we return `true` now at my suggestion) when calling `Height::satisfies_lock_time(time_based_lock_time)` is possibly confusing. We could return an error but if `satisfies_lock_time` is a trait method then I really don't like this because then all the sane implementors will have to have an infallible error type which would suck with our MSRV.

I'm okay with the current API but my alternative would be for `Height` and `Time` to have their own, non-trait, `satisfies_lock_time` methods that return a `Result` and to leave the other 3 impls alone.

tcharding commented:
I added separate functions [back in] `LockTime::is_satisfied_by_height` and `LockTime::is_satisfied_by_time` that return a `Result`. Implemented them in bit of a roundabout way by calling `Height::satisfies_lock_time` which also returns a result (i.e., is a method not a trait method).

This results in a much cleaner API I found, as show by the examples in this version.

tcharding commented:
Woops, I just squashed the `transaction::RelativeLockTimeError` -> `relative::Error` change into the last commit. I think we all know whats going on now, so leaving it as such. Holla if you want it split out.

tcharding commented:
Rust question please: I've changed all the methods to take ownership of `self`. I believe we have discussed this before but I find myself again not knowing when to take ownership of self. I have been defaulting to only ever taking owneship if its needed in the method body. For `Copy` types should methods always take ownership?

Kixunil commented:
@tcharding `Copy` types are more convenient when taken by value. @dr-orlovsky had a good point they are convenient in combinators (`iter.map(Type::some_fn)`). Although maybe those returning `bool` are more convenient to take by reference because `filter` takes a reference (OTOH if you want to map iter to bools it sucks but hopefully less likely?)

Kixunil commented:
> We could return an error but if `satisfies_lock_time` is a trait method then I really don't like this because then all the sane implementors will have to have an infallible error type which would suck with our MSRV.

The annoying part is calling `.unwrap_or_else(|e| match e {})`, indeed. An alternative suggestion:

```rust
pub trait SatisfiesRelativeLockTime: sealed::SatisfiesRelativeLockTime {
    type Error;

    fn satisfies_relative_lock_time(self, lock_time: relative::LockTime) -> Result<bool, Self::Error>;
}

// impl for `(Height, Time)`, `(Time, Height)` with `Error = Infallible`
// impl for `Height`, `Time, `Sequence` with `Error = UnitMismatchError`

impl relative::LockTime {
    pub fn is_satisfied_by<T: SatisfiesRelativeLockTime<Error=Infallible>>(self, satisified_by: T) -> bool {
        satisfied_by.satisfies_relative_lock_time(self).unwrap_or_else(|never| match never {})
    }

    pub fn try_is_satisfied_by<T: SatisfiesRelativeLockTime>(self, satisified_by: T) -> Result<bool, T::Error> {
        satisfied_by.satisfies_relative_lock_time(self)
    }
}
```

apoelstra reviewed (dismissed):
ACK b55fe3d89abb6b510c08aafed777af91a55111a9

apoelstra commented:
This looks good to me, though I'm not convinced that it's simpler than the equivalent code where we drop the traits and just (repetitively) inline all the impls, returning `Result` or not as appropriate.

tcharding commented:
Ha, after adding the result returning ones yesterday that thought did cross my mind. Glad you mentioned it. I removed the traits, simple code is goal. It was an educational exercise however.

apoelstra commented:
I think this looks great! Maybe @Kixunil will disagree since we moved away from his suggested trait-based style.

apoelstra reviewed (approved):
ACK da95c3a489c18399d6c3170a0c8af0b5fa8d7cf9

tcharding commented:
> I think this looks great! Maybe @Kixunil will disagree since we moved away from his suggested trait-based style.

I don't think so since when that was suggested we didn't realize we wanted to return an error sometimes. @Kixunil, you cool with this as is?

sanket1729 commented on src/blockdata/locktime/relative.rs:
```diff
@@ -0,0 +1,289 @@
+    ///
+    /// Will return an error if the input cannot be encoded in 16 bits.
+    #[inline]
+    pub fn from_seconds_ceil(seconds: u32) -> Result<Self, Error> {
```
nit: Can this be an `Option<T>` instead? There is only one variant that can happen here. Most arithmetic operations in rust return an Option when the error is only possible before as overflow. 
I would have suggested a dedicated error type, but it would be some work and a lot of code. 

This would be annoying to `unwrap` because we not know all error variants

tcharding replied:
hmmm, a few considerations
1. There is an equivalent method on `Sequence` so we should do the same for both
2. It was a dedicated error type until I added the other variants
3. Option has less information because the error returns the input value, is that useful? I don't know really (@Kixunil is the resident "useful error message" bloke, what do you think man?)

I can do either
1. split the error into two different types
2. Use an `Option` as suggested

I tend to agree that `2` is the best

tcharding replied:
Either way, I'll do it as a follow up please - this one has been running for long enough now.

Kixunil replied:
I personally prefer custom error types to `Option` even at the cost of added code. It provides more information when people blindly `?` it. (and `.ok()` is possible if they want to ignore)

A special type would be better indeed.

Kixunil replied:
To expand on this a bit, in my experience sprinkling `.ok_or{_else}` or `map_err` everywhere is quite annoying and not doing it will bite you when function deep in the call stack returns error (`None`) and you don't even know which one it was. For instance this is very common when people `?` `std::io::Error`.

apoelstra replied:
@sanket1729 could you raise an issue about this, if you feel strongly enough? I am going to merge this PR.

tcharding replied:
Furthering this discussion, I'm interested to learn if our method of having a single enum error type in each module (typically) is the correct way to go about things, alternatives (off the top of my head) are:
- An individual error struct for every error
- An enum for related errors
- An enum error per "some other semantic block" e.g. all of the networking code

Kixunil replied:
My initial thoughts: for types convertible from byte arrays and such, parsing errors should be separate from other conversion errors - different layers-different errors
Functions that are very often called together can share error.
Everything else should have its own.

Splitting into crates will force us to at least split errors from independent parts.

sanket1729 replied:
Don't feel so strongly about it now. My workflow for detecting it is to try using it on a type that I know it should succeed on and match on error types before returning. I will raise issue/PR later about it

sanket1729 reviewed (approved):
ACK da95c3a489c18399d6c3170a0c8af0b5fa8d7cf9. Left a minor nit

Kixunil commented:
I don't mind avoiding the trait, it was just an idea.
